File: src/main/variations/Variation.java
package main.variations;

import main.Entity;
import main.buttons.Element;
import main.buttons.Pack;
import main.variations.appearances.Appearance;
import org.apache.commons.lang3.tuple.ImmutablePair;
import processing.core.PImage;
import processing.data.JSONArray;
import processing.data.JSONObject;

import java.util.ArrayList;

public abstract class Variation extends Entity {

    JSONObject json;
    Element element;

    Variation(JSONObject json, Element element) {
        this.json = json;
        this.element = element;
        //can't load images in the constructor because it won't be in the image threading process
    }

    public PImage getImage() {
        PImage image = this.getAppearance().getImage();
        return image == null ? this.element.getImage() : image;
    }

    public abstract void loadImages();

    public Element getElement() {
        return this.element;
    }

    public String getName() {
        return this.getAppearance().getName();
    }

    ArrayList<Appearance> loadAppearances() {
        ArrayList<Appearance> list = new ArrayList<>();
        JSONArray textures = this.json.getJSONArray("textures");
        for (int i = 0; i < textures.size(); i++) {
            Object object = textures.get(i);
            if (object instanceof String) {
                JSONObject json = new JSONObject();
                json.put("texture", object);
                list.add(Appearance.getAppearance(this, json));
            } else {
                list.add(Appearance.getAppearance(this, (JSONObject) object));
            }
        }
        return list;
    }

    public abstract Appearance getAppearance();

    public abstract ArrayList<ImmutablePair<PImage, String>> getPairs();

    public static Variation getVariation(JSONObject json, Element element, Pack pack) {
        switch (json.getString("type")) {
            case "random":
                return new RandomVariation(json, element);
            case "combo":
                return new ComboVariation(json, element);
            case "month":
                return new MonthVariation(json, element);
            case "week":
                return new WeekVariation(json, element);
            case "inherit":
                return new InheritVariation(json, element, pack);
            case "animation":
                return new AnimationVariation(json, element);
            default:
                return null;
        }
    }

}

File: src/main/variations/Variation.kt
package main.variations

import main.Element
import main.Entity
import main.buttons.Pack
import main.variations.appearances.Appearance
import main.variations.appearances.Appearance.Companion.getAppearance
import processing.core.PImage
import processing.data.JSONObject
import java.util.*

abstract class Variation  //can't load images in the constructor because it won't be in the image threading process
internal constructor(val json: JSONObject, val element: Element) : Entity() {
    open fun getImage(): PImage = getAppearance().getImage()

    abstract fun loadImages()

    open fun getName(): String = getAppearance().getName()

    fun loadAppearances(): ArrayList<Appearance> {
        val list = ArrayList<Appearance>()
        val textures = json.getJSONArray("textures")
        for (i in 0 until textures.size()) {
            val `object` = textures[i]
            if (`object` is String) {
                val json = JSONObject()
                json.put("texture", `object`)
                list.add(getAppearance(this, json))
            } else {
                list.add(getAppearance(this, (`object` as JSONObject)))
            }
        }
        return list
    }

    abstract fun getAppearance(): Appearance

    abstract fun getPairs(): List<Pair<PImage, String>>

    companion object {
        @JvmStatic
        fun getVariation(json: JSONObject, element: Element, pack: Pack): Variation? {
            return when (json.getString("type")) {
                "random" -> RandomVariation(json, element)
                "combo" -> ComboVariation(json, element)
                "month" -> MonthVariation(json, element)
                "week" -> WeekVariation(json, element)
                "inherit" -> InheritVariation(json, element, pack)
                "animation" -> AnimationVariation(json, element)
                else -> null
            }
        }
    }
}